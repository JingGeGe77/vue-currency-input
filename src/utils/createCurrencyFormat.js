const createCurrencyFormat = (numberFormat) => {
  const ps = numberFormat.format(123456)
  const ns = numberFormat.format(-1)
  const minimumFractionDigits = numberFormat.resolvedOptions().minimumFractionDigits || 0
  const maximumFractionDigits = numberFormat.resolvedOptions().maximumFractionDigits || 0
  const decimalSymbol = minimumFractionDigits > 0 ? ps.substr(ps.indexOf('6') + 1, 1) : null
  const prefix = ps.substring(0, ps.indexOf('1'))
  const negativePrefix = ns.substring(0, ns.indexOf('1'))
  const suffix = ps.substring(ps.lastIndexOf(minimumFractionDigits > 0 ? '0' : '6') + 1)
  const groupingSymbol = ps.substr(ps.indexOf('3') + 1, 1)

  return {
    prefix,
    negativePrefix,
    suffix,
    groupingSymbol,
    decimalSymbol,
    minimumFractionDigits,
    maximumFractionDigits
  }
}

export default ({ locale, currency, precision, autoDecimalMode, valueAsInteger }) => {
  let minimumFractionDigits = 2
  let maximumFractionDigits = 2
  if (typeof precision === 'number') {
    minimumFractionDigits = maximumFractionDigits = precision
  } else if (typeof precision === 'object' && !autoDecimalMode && !valueAsInteger) {
    minimumFractionDigits = precision.min || 0
    maximumFractionDigits = precision.max !== undefined ? precision.max : 20
  }
  if (currency == null) {
    return createCurrencyFormat(new Intl.NumberFormat(locale, { minimumFractionDigits, maximumFractionDigits }))
  } else if (typeof currency === 'object') {
    return {
      ...createCurrencyFormat(new Intl.NumberFormat(locale, { minimumFractionDigits, maximumFractionDigits })),
      prefix: currency.prefix || '',
      negativePrefix: `-${currency.prefix || ''}`,
      suffix: currency.suffix || ''
    }
  } else {
    const currencyFormat = createCurrencyFormat(new Intl.NumberFormat(locale, { currency, style: 'currency' }))
    if (precision !== undefined) {
      if (currencyFormat.minimumFractionDigits > 0) {
        currencyFormat.minimumFractionDigits = minimumFractionDigits
      }
      if (currencyFormat.maximumFractionDigits > 0) {
        currencyFormat.maximumFractionDigits = maximumFractionDigits
      }
    }
    return currencyFormat
  }
}
