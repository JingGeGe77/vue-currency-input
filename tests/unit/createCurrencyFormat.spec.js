import Intl from 'intl'
import createCurrencyFormat from '../../src/utils/createCurrencyFormat'

global.Intl = Intl

describe('createCurrencyFormat', () => {
  it('de-DE_EUR', () => {
    expect(createCurrencyFormat({ locale: 'de-DE', currency: 'EUR' })).toMatchSnapshot()
  })

  it('es-ES_EUR', () => {
    expect(createCurrencyFormat({ locale: 'es-ES', currency: 'EUR' })).toMatchSnapshot()
  })

  it('nl-NL_EUR', () => {
    expect(createCurrencyFormat({ locale: 'nl-NL', currency: 'USD' })).toMatchSnapshot()
  })

  it('en-US_USD', () => {
    expect(createCurrencyFormat({ locale: 'en-US', currency: 'USD' })).toMatchSnapshot()
  })

  it('zh_CNY', () => {
    expect(createCurrencyFormat({ locale: 'zh', currency: 'CNY' })).toMatchSnapshot()
  })

  it('en-GB_GBP', () => {
    expect(createCurrencyFormat({ locale: 'en-GB', currency: 'GBP' })).toMatchSnapshot()
  })

  it('en-IN_INR', () => {
    expect(createCurrencyFormat({ locale: 'en-GB', currency: 'INR' })).toMatchSnapshot()
  })

  it('pt_BRL', () => {
    expect(createCurrencyFormat({ locale: 'pt', currency: 'BRL' })).toMatchSnapshot()
  })

  it('ja_JPY', () => {
    expect(createCurrencyFormat({ locale: 'ja', currency: 'JPY' })).toMatchSnapshot()
  })
})
