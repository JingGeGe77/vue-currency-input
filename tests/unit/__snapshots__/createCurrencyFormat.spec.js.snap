// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCurrencyFormat custom precision 1`] = `
Object {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`createCurrencyFormat custom precision 2`] = `
Object {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`createCurrencyFormat custom precision 3`] = `
Object {
  "decimalSymbol": null,
  "groupingSymbol": ",",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "negativePrefix": "-￥",
  "prefix": "￥",
  "suffix": "",
}
`;

exports[`createCurrencyFormat custom prefix/suffix 1`] = `
Object {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "-₿ ",
  "prefix": "₿ ",
  "suffix": "",
}
`;

exports[`createCurrencyFormat custom prefix/suffix 2`] = `
Object {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " Euro",
}
`;

exports[`createCurrencyFormat custom prefix/suffix 3`] = `
Object {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "-",
  "prefix": "",
  "suffix": "",
}
`;

exports[`createCurrencyFormat i18n de-DE_EUR 1`] = `
Object {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`createCurrencyFormat i18n en-GB_GBP 1`] = `
Object {
  "decimalSymbol": ".",
  "groupingSymbol": ",",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "-£",
  "prefix": "£",
  "suffix": "",
}
`;

exports[`createCurrencyFormat i18n en-IN_INR 1`] = `
Object {
  "decimalSymbol": ".",
  "groupingSymbol": ",",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "-₹",
  "prefix": "₹",
  "suffix": "",
}
`;

exports[`createCurrencyFormat i18n en-US_USD 1`] = `
Object {
  "decimalSymbol": ".",
  "groupingSymbol": ",",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "-$",
  "prefix": "$",
  "suffix": "",
}
`;

exports[`createCurrencyFormat i18n es-ES_EUR 1`] = `
Object {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "-",
  "prefix": "",
  "suffix": " €",
}
`;

exports[`createCurrencyFormat i18n ja_JPY 1`] = `
Object {
  "decimalSymbol": null,
  "groupingSymbol": ",",
  "maximumFractionDigits": 0,
  "minimumFractionDigits": 0,
  "negativePrefix": "-￥",
  "prefix": "￥",
  "suffix": "",
}
`;

exports[`createCurrencyFormat i18n nl-NL_EUR 1`] = `
Object {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "€ -",
  "prefix": "€ ",
  "suffix": "",
}
`;

exports[`createCurrencyFormat i18n pt_BRL 1`] = `
Object {
  "decimalSymbol": ",",
  "groupingSymbol": ".",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "-R$",
  "prefix": "R$",
  "suffix": "",
}
`;

exports[`createCurrencyFormat i18n zh_CNY 1`] = `
Object {
  "decimalSymbol": ".",
  "groupingSymbol": ",",
  "maximumFractionDigits": 2,
  "minimumFractionDigits": 2,
  "negativePrefix": "-￥",
  "prefix": "￥",
  "suffix": "",
}
`;
